File: apps/android/app/src/main/java/dev/flextranslate/foundation/MilmmtMtProvider.kt
package dev.flextranslate.foundation

import android.util.Log
import java.io.File

/**
 * Реальный локальный перевод через MiLMMT-46-4B (архитектура Gemma-3), Q6_K GGUF, поверх вендоренного
 * llama.cpp ([LlamaCppBridge]). Реализует [TranslationProvider]. Это QUALITY-тир WS4 — пользователь
 * выбирает его против сбалансированного M2M-100 и облака.
 *
 * Честность (G005/WS4 A2): любой не-null [TranslationResult.text] — НАСТОЯЩИЙ вывод модели, раскодированный
 * llama.cpp. Ничего не выдумываем: нет модели, недоступен нативный runtime, упал decode — отдаём честную
 * [TranslationResult.unsupportedReason] (гейт), а не фейковый текст.
 *
 * Модель и контекст создаём лениво при первом вызове и переиспользуем — загрузка 4B GGUF дорогая (секунды).
 * 4B LLM на мобиле медленный (секунды на перевод); для quality-тира это нормально и относится к бенчмарку/гейту
 * (WS6), а не к корректности. Тяжёлая загрузка/генерация идёт не в главном потоке — вызывающий
 * ([LiveSessionState.translateFinal]) уже дёргает [translate] на воркер-потоке.
 *
 * Формат промпта — из карточки модели (xiaomi-research/MiLMMT-46-4B-v0.1): completion-style инструкция
 * с полными названиями языков источника и цели:
 *   `Translate this from <Source> to <Target>:\n<Source>: <text>\n<Target>:`
 */
class MilmmtMtProvider(
    private val spec: MtModelSpec.Gguf,
    private val modelDir: File,
    private val bridge: LlamaCppBridge = LlamaCppBridge,
) : TranslationProvider {

    override val providerId: String = "milmmt:${spec.modelId}"

    private sealed interface State {
        data object Uninitialized : State
        data class Ready(val handle: Long) : State
        data object MissingModel : State
        data object RuntimeUnavailable : State
        data class Failed(val reason: String) : State
    }

    private var state: State = State.Uninitialized

    override fun translate(text: String, languagePair: String, deviceTier: String): TranslationResult {
        val trimmed = text.trim()
        if (trimmed.isEmpty()) return TranslationResult(text = null, unsupportedReason = null)

        val direction = MilmmtDirection.parse(languagePair)
            ?: return TranslationResult(
                text = null,
                unsupportedReason = "пара $languagePair не поддерживается этой моделью",
            )

        return when (val ready = ensureModel()) {
            is State.Ready -> runTranslate(ready.handle, trimmed, direction)
            State.MissingModel -> TranslationResult(
                text = null,
                unsupportedReason = "MT-модель ${spec.modelId} не установлена",
            )
            State.RuntimeUnavailable -> TranslationResult(
                text = null,
                unsupportedReason = "MiLMMT runtime ещё не установлен",
            )
            is State.Failed -> TranslationResult(
                text = null,
                unsupportedReason = "MT-движок недоступен: ${ready.reason}",
            )
            State.Uninitialized -> TranslationResult(
                text = null,
                unsupportedReason = "MT-движок не инициализирован",
            )
        }
    }

    private fun runTranslate(handle: Long, text: String, direction: MilmmtDirection): TranslationResult {
        val prompt = buildPrompt(text, direction)
        val output = bridge.generate(handle, prompt, MAX_NEW_TOKENS)
            ?: return TranslationResult(
                text = null,
                unsupportedReason = "ошибка генерации llama.cpp",
            )
        val cleaned = output.trim()
        if (cleaned.isEmpty()) {
            return TranslationResult(text = null, unsupportedReason = "пустой ответ модели")
        }
        return TranslationResult(text = cleaned, unsupportedReason = null)
    }

    /** Completion-style промпт из карточки MiLMMT; полные английские названия языков. */
    private fun buildPrompt(text: String, direction: MilmmtDirection): String =
        "Translate this from ${direction.sourceName} to ${direction.targetName}:\n" +
            "${direction.sourceName}: $text\n" +
            "${direction.targetName}:"

    /** Честная проверка готовности без полной загрузки модели — когда файла просто нет. */
    fun isModelInstalled(): Boolean = spec.isInstalled(modelDir)

    private fun ensureModel(): State {
        val current = state
        if (current !is State.Uninitialized) return current
        if (!bridge.isAvailable) {
            state = State.RuntimeUnavailable
            return state
        }
        if (!spec.isInstalled(modelDir)) {
            state = State.MissingModel
            return state
        }
        val ggufPath = File(modelDir, spec.gguf).absolutePath
        val handle = bridge.load(ggufPath, DEFAULT_THREADS, DEFAULT_CTX)
        state = if (handle != 0L) {
            Log.i(TAG, "MiLMMT loaded: $ggufPath")
            State.Ready(handle)
        } else {
            State.Failed("llama.cpp model load returned 0")
        }
        return state
    }

    override fun close() {
        (state as? State.Ready)?.let { bridge.free(it.handle) }
        state = State.Uninitialized
    }

    private companion object {
        const val TAG = "MilmmtMtProvider"
        // 4B на мобиле: держим контекст скромным (реплики в диалоге короткие) ради памяти и скорости.
        const val DEFAULT_CTX = 1024
        const val DEFAULT_THREADS = 4
        const val MAX_NEW_TOKENS = 128
    }
}

/**
 * Разобранное направление `"src->tgt"` в пределах демо-набора языков. Несёт полные английские названия,
 * которые требует промпт MiLMMT (напр. `"zh"` → `"Chinese (Simplified)"`).
 */
data class MilmmtDirection(val source: String, val target: String, val sourceName: String, val targetName: String) {
    companion object {
        private val LANGUAGE_NAMES = mapOf(
            "ru" to "Russian",
            "en" to "English",
            "zh" to "Chinese (Simplified)",
        )

        /** Разбирает `"en->ru"`; null если формат битый или язык вне демо-набора RU/EN/ZH. */
        fun parse(languagePair: String): MilmmtDirection? {
            val parts = languagePair.split("->", "-", "_", limit = 2)
            if (parts.size != 2) return null
            val source = parts[0].trim().lowercase()
            val target = parts[1].trim().lowercase()
            val sourceName = LANGUAGE_NAMES[source] ?: return null
            val targetName = LANGUAGE_NAMES[target] ?: return null
            if (source == target) return null
            return MilmmtDirection(source, target, sourceName, targetName)
        }
    }
}

File: apps/ios/FlexTranslate/Sources/FlexTranslate/MilmmtMtProvider.swift
import Foundation

// On-device перевод через MiLMMT-46-4B (архитектура Gemma-3) в формате Q6_K GGUF,
// крутится на вендоренном llama.cpp xcframework. Это QUALITY-уровень: пользователь
// может выбрать его наряду с M2M-100 (баланс) и Gemini (облако).
//
// Дисциплина A2: любой непустой TranslationResult.text — это РЕАЛЬНЫЙ вывод модели,
// раскодированный llama.cpp. Ничего не выдумываем: нет модели, недоступен фреймворк
// или сорвалось декодирование — честно отдаём unsupportedReason, а не фейковый текст.
//
// Модель и контекст создаём лениво при первом вызове и переиспользуем — загрузка 4B
// GGUF дорогая, это секунды. 4B-модель на телефоне/симуляторе медленная (секунды на
// перевод), но для quality-уровня это нормально.
// Тяжёлая работа идёт вне main actor — вызывающий уже уводит нас в фоновый поток через
// Thread.detachNewThread в LiveSessionModel.
//
// Формат промпта взят из карточки модели (xiaomi-research/MiLMMT-46-4B-v0.1):
//   Translate this from <Source> to <Target>:\n<Source>: <text>\n<Target>:
// Точная калька с Android-версии MilmmtMtProvider.
final class MilmmtMtProvider: TranslationProvider {
    let providerId: String

    private let spec: MtModelSpec.GgufConfig
    private let modelDir: URL

    private enum State {
        case uninitialized
        case ready(OpaquePointer)
        case missingModel
        case failed(String)
    }

    private let stateLock = NSLock()
    private var _state: State = .uninitialized

    init(spec: MtModelSpec.GgufConfig, modelDir: URL) {
        self.spec = spec
        self.modelDir = modelDir
        self.providerId = "milmmt:\(spec.modelId)"
    }

    // MARK: - TranslationProvider

    func translate(text: String, languagePair: String, deviceTier: String) -> TranslationResult {
        let trimmed = text.trimmingCharacters(in: .whitespaces)
        if trimmed.isEmpty {
            return TranslationResult(text: nil, unsupportedReason: nil)
        }

        guard let direction = MilmmtDirection.parse(languagePair) else {
            return TranslationResult(
                text: nil,
                unsupportedReason: "пара \(languagePair) не поддерживается этой моделью"
            )
        }

        switch ensureModel() {
        case .ready(let handle):
            return runTranslate(handle: handle, text: trimmed, direction: direction)
        case .missingModel:
            return TranslationResult(
                text: nil,
                unsupportedReason: "MT-модель \(spec.modelId) не установлена"
            )
        case .failed(let reason):
            return TranslationResult(
                text: nil,
                unsupportedReason: "MT-движок недоступен: \(reason)"
            )
        case .uninitialized:
            return TranslationResult(
                text: nil,
                unsupportedReason: "MT-движок не инициализирован"
            )
        }
    }

    var isModelInstalled: Bool {
        MtModelSpec.gguf(spec).isInstalled(in: modelDir)
    }

    func close() {
        stateLock.lock()
        let current = _state
        _state = .uninitialized
        stateLock.unlock()
        if case .ready(let handle) = current {
            LlamaCppBridge.free(handle: handle)
        }
    }

    // MARK: - Private

    private func runTranslate(handle: OpaquePointer, text: String, direction: MilmmtDirection) -> TranslationResult {
        let prompt = buildPrompt(text: text, direction: direction)
        guard let output = LlamaCppBridge.generate(
            handle: handle,
            prompt: prompt,
            maxNewTokens: LlamaCppBridge.maxNewTokens
        ) else {
            return TranslationResult(
                text: nil,
                unsupportedReason: "ошибка генерации llama.cpp"
            )
        }
        let cleaned = output.trimmingCharacters(in: .whitespacesAndNewlines)
        if cleaned.isEmpty {
            return TranslationResult(text: nil, unsupportedReason: "пустой ответ модели")
        }
        return TranslationResult(text: cleaned, unsupportedReason: nil)
    }

    // Промпт в completion-стиле из карточки MiLMMT, языки пишем полными английскими названиями.
    // Точная калька с Android MilmmtMtProvider.buildPrompt.
    private func buildPrompt(text: String, direction: MilmmtDirection) -> String {
        "Translate this from \(direction.sourceName) to \(direction.targetName):\n" +
        "\(direction.sourceName): \(text)\n" +
        "\(direction.targetName):"
    }

    private func ensureModel() -> State {
        stateLock.lock()
        let current = _state
        stateLock.unlock()

        if case .uninitialized = current {} else { return current }

        // Сначала проверяем, что файл модели на месте — грузить впустую дорого.
        if !MtModelSpec.gguf(spec).isInstalled(in: modelDir) {
            stateLock.lock()
            _state = .missingModel
            stateLock.unlock()
            return .missingModel
        }

        let ggufPath = modelDir.appendingPathComponent(spec.gguf).path
        if let handle = LlamaCppBridge.load(
            path: ggufPath,
            nThreads: LlamaCppBridge.defaultThreads,
            nCtx: LlamaCppBridge.defaultCtx
        ) {
            let next = State.ready(handle)
            stateLock.lock()
            _state = next
            stateLock.unlock()
            return next
        } else {
            let next = State.failed("llama.cpp model load returned nil")
            stateLock.lock()
            _state = next
            stateLock.unlock()
            return next
        }
    }
}

// MARK: - Direction parsing

// Распарсенное направление "src->tgt" в рамках демо-языков. Хранит полные английские
// названия, которые нужны промпту MiLMMT.
// Точная калька с Android MilmmtDirection.
struct MilmmtDirection: Equatable {
    let source: String
    let target: String
    let sourceName: String
    let targetName: String

    private static let languageNames: [String: String] = [
        "ru": "Russian",
        "en": "English",
        "zh": "Chinese (Simplified)",
    ]

    // Разбирает "en->ru"; nil, если формат битый или язык вне демо-набора RU/EN/ZH.
    static func parse(_ languagePair: String) -> MilmmtDirection? {
        let separators = ["->", "-", "_"]
        var parts: [String]?
        for sep in separators {
            let split = languagePair.components(separatedBy: sep)
            if split.count == 2 {
                parts = split
                break
            }
        }
        guard let parts else { return nil }
        let source = parts[0].trimmingCharacters(in: .whitespaces).lowercased()
        let target = parts[1].trimmingCharacters(in: .whitespaces).lowercased()
        guard
            let sourceName = languageNames[source],
            let targetName = languageNames[target],
            source != target
        else { return nil }
        return MilmmtDirection(source: source, target: target,
                               sourceName: sourceName, targetName: targetName)
    }
}
